### PR TITLE
UCF101 dataset more *loading* efficient

### PR DESCRIPTION
**Pull request by Guillem96**
Now when creating UCF101 dataset, you specify both the fold ({1, 2, 3}) and the split (train or test). Due to the current implementation, the video clip generation is agnostic to the selected fold and split, meaning that the `VideoClips` helper class processes all the videos first and afterwards, according to the selected split and fold, the clips are filtered.

This is very inefficient.  For example, if you want to load only the test set, you have to go through around 900 videos while test set only contains around 350 videos. 

To solve this I:

1. Generate the dataset samples with `make_dataset` independently from split and fold.
2. Obtain the videos' path list from annotation file corresponding to the selected split and fold
3. Filter the dataset samples obtained in the step 1, so we only keep the samples that belong to the desired split and fold
4. Process the videos with the `VideoClips` object.

### CODE REVIEW

File: torchvision/datasets/ucf101.py [hunk 1]
```diff
@@ -71,35 +81,30 @@ def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
             _video_min_dimension=_video_min_dimension,
             _audio_samples=_audio_samples,
         )
-        self.video_clips_metadata = video_clips.metadata
-        self.indices = self._select_fold(video_list, annotation_path, fold, train)
-        self.video_clips = video_clips.subset(self.indices)
-        self.transform = transform
+        self.video_clips_metadata = self.video_clips.metadata
```
Comment (fmassa) [inline torchvision/datasets/ucf101.py:84]: This is unfortunately a BC-breaking change as other downstream projects rely on this behavior, see [ClassyVision](https://github.com/facebookresearch/ClassyVision/blob/7632f2a3f9ff8b753d350208659e97dbf90f78a7/classy_vision/dataset/classy_ucf101.py#L102) for example.

The original thinking behind this approach was that one could cache the metadata and re-use it over different dataset invocations, so that the creation time would be amortized.

It could indeed be possible to from the beginning create separate metadata for each fold, but now that it's been done as is we unfortunately to keep it for backwards-compatibility reasons.

**Comment by vreis** [comment]
@stephenyan1231: I'll defer to you on this one. Seems like a good change, and in general I don't mind breaking changes. But I don't know if we rely on this behavior in subtle ways.

**Comment by Guillem96** [comment]
Any news on this?

**Comment by yassineAlouini** [comment]
Thanks @Guillem96 for this contribution and sorry for taking that long to get an update.

As you might have read, the dataset API is being reworked and datasets are being moved to the new API (as detailed [here](https://github.com/pytorch/vision/issues/5336)). 

Not yet sure what is the best course of action but here are two possible paths:

* add some tests to the current implementation and update it to make sure that it is backward compatible (maybe this will be hard/impossible to do?) then migrate to the new dataset API. The migration can be done by someone else of course and due credit will be given to you @Guillem96. :ok_hand: 
* first migrate the dataset to the new dataset API (again can be done by someone else) and then integrate the suggested code while making sure to keep the code backward compatible (if that's possible).

In the two paths, I would suggest having two separate PRs so that the work is easier to review.

What works best for you @Guillem96? Feel free if you have questions and/or other suggestions.

Any other suggestions @pmeier on this?

**Comment by pmeier** [comment]
We are currently trying to figure out the best way to do this in #5422 cc @bjuncek. The new design will not re-use the `VideoClips` object as we will not be able to index the whole dataset upfront. Thus, I don't think we need to keep this open. @Guillem96 are you ok with closing this and following the development in the new PR?

**Comment by Guillem96** [comment]
@pmeier fine with me! Thanks for your comments👌🏼
